### ISSUE TITLE
btrbk v0.28.0 crashes if deep parent/child relations exceed depth=100

### ISSUE THREAD

**Issue by digint** [digint/btrbk#279]
Perl hates recursions, and dies while searching for related subvolumes if recursion depth = 100:
```
Stack Trace:
----------------------------------------
Deep recursion on subroutine "main::_push_related_children" at /usr/sbin/btrbk line 3164.
 at /usr/sbin/btrbk line 294.
	main::__ANON__("Deep recursion on subroutine \"main::_push_related_children\" a"...) called at /usr/sbin/btrbk line 3164
	main::_push_related_children(HASH(0x56117423b400), "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx0", HASH(0x561173c23528), ARRAY(0x561173cc2e88), undef, 99) called at /usr/sbin/btrbk line 3164
	main::_push_related_children(HASH(0x56117423b400), "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx1", HASH(0x561173c23528), ARRAY(0x561173cc2e88), undef, 98) called at /usr/sbin/btrbk line 3164
	[...]
	main::_push_related_children(HASH(0x56117423b400), "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx1", HASH(0x561173c23528), ARRAY(0x561173cc2e88), undef, 0) called at /usr/sbin/btrbk line 3164
	main::get_related_readonly_nodes(HASH(0x5611748bf8a0)) called at /usr/sbin/btrbk line 3241
```
    
Unfortunately this happens before (hardcoded) [abort condition distance=256](https://github.com/digint/btrbk/blob/v0.28.0/btrbk#L3138).

(Originally reported by A.H. via email)

**Comment by digint** [comment]
Preliminary bugfix (mostly untested) is available in "fix-related-search" branch, commit cd1d9d8ed40d2f8654ee3ab277884c8dcd31316c

**Comment by alexh-name** [comment]
Hi, that was me. Just checking in to offer further testing.

**Comment by alexh-name** [comment]
cd1d9d8ed40d2f8654ee3ab277884c8dcd31316c works, thanks!
(Expectantly running into `WARNING: Parent chain exceeds depth=256, ignoring related parents of ...`)

**Comment by digint** [comment]
Pushed fix to master, with some amendments:

 * 37b0bd34773fbf21862fe5de343c7b962076faeb : btrbk: bugfix: make get_related_readonly_nodes non-recursive
 * b62bb3b1a8d77def6cdfe5ccf221758a50ed5719 : btrbk: get_related_readonly_nodes: vastly increase abort_distance

The warnings should be gone, except if you have deep parent/child relationship on the source. On "btrbk archive", no warnings are given as deep relations are not relevant here (only close relations are needed for picking best parent for send-receive).

**Comment by alexh-name** [comment]
Can confirm that newest version works fine and warnings are gone. Thanks!

**Comment by blinkeye** [comment]
I had the same issue with `v0.28.0`, tried the current master https://github.com/digint/btrbk/commit/b62bb3b1a8d77def6cdfe5ccf221758a50ed5719 and could successfully archive my backups. Thanks a lot!

**Comment by digint** [comment]
Fix included in [btrbk-0.28.1](/digint/btrbk/releases/tag/v0.28.1).
